pkg/controller/certificates/approver/groupapprove.go

```diff
@@ -0,0 +1,181 @@
+	authorizationapi "k8s.io/kubernetes/pkg/apis/authorization/v1beta1"
+	capi "k8s.io/kubernetes/pkg/apis/certificates/v1beta1"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	certificatesinformers "k8s.io/kubernetes/pkg/client/informers/informers_generated/externalversions/certificates/v1beta1"
+	"k8s.io/kubernetes/pkg/controller/certificates"
+)
+
+const (
+	selfNodeClientCertAccess = "certificatessigningrequests/selfnodeclientcert"
+	nodeClientCertAccess     = "certificatessigningrequests/nodeclientcert"
+	nodeServerCertAccess     = "certificatessigningrequests/nodeservercert"
+)
+
+func NewCSRApprovingController(client clientset.Interface, csrInformer certificatesinformers.CertificateSigningRequestInformer) (*certificates.CertificateController, error) {
+	approver := &groupApprover{
+		client: client,
+	}
+	return certificates.NewCertificateController(
+		client,
+		csrInformer,
+		approver.handle,
+	)
+}
+
+// groupApprover implements AutoApprover for signing Kubelet certificates.
+type groupApprover struct {
+	client clientset.Interface
+}
+
+func (ga *groupApprover) handle(csr *capi.CertificateSigningRequest) error {
+	// short-circuit if we're already approved or denied
+	if approved, denied := certificates.GetCertApprovalCondition(&csr.Status); approved || denied {
+		return nil
+	}
+	csr, err := ga.autoApprove(csr)
+	if err != nil {
+		return fmt.Errorf("error auto approving csr: %v", err)
+	}
+	_, err = ga.client.Certificates().CertificateSigningRequests().UpdateApproval(csr)
+	if err != nil {
+		return fmt.Errorf("error updating approval for csr: %v", err)
+	}
+	return nil
+}
+
+func (cc *groupApprover) autoApprove(csr *capi.CertificateSigningRequest) (*capi.CertificateSigningRequest, error) {
+	x509cr, err := capi.ParseCSR(csr)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("unable to parse csr %q: %v", csr.Name, err))
+		return csr, nil
+	}
+	if !reflect.DeepEqual([]string{"system:nodes"}, x509cr.Subject.Organization) {
+		return csr, nil
+	}
+	if len(x509cr.DNSNames)+len(x509cr.EmailAddresses)+len(x509cr.IPAddresses) != 0 {
+		return csr, nil
+	}
+	if hasExactUsages(csr, kubeletClientUsages) {
+		if !strings.HasPrefix(x509cr.Subject.CommonName, "system:node:") {
+			return csr, nil
+		}
+		// check kubelet up for client cert renewal
+		if csr.Spec.Username == x509cr.Subject.CommonName {
+			approve, err := cc.doAccessReview(csr, selfNodeClientCertAccess)
+			if err != nil {
+				return nil, err
+			}
+			if approve {
+				approveWithMessage(csr, "Auto approving of all kubelet client certificate rotation after SAR.")
+				return csr, nil
+			}
+		}
+		// check kubelet client cret bootstrapper
+		approve, err := cc.doAccessReview(csr, nodeClientCertAccess)
+		if err != nil {
+			return nil, err
+		}
+		if approve {
+			approveWithMessage(csr, "Auto approving of all kubelet client certificate rotation after SAR.")
+			return csr, nil
+		}
+	}
+	if hasExactUsages(csr, kubeletServerUsages) {
```

Comment:
have you thought about how we'd verify what hostnames and subjectaltnames a kubelet is allowed to ask for in its serving cert?